src/FFNetAgent.cpp

```diff
@@ -0,0 +1,365 @@
+    {
+        return m_do_write_batch;
+    }
+
+    // Read signals from the platform and calculate samples to be sent up
+    void FFNetAgent::sample_platform(std::vector<double> &out_sample)
+    {
+        ++m_sample;
+
+        for (auto &kv : m_net_map) {
+            kv.second->sample();
+        }
+    }
+
+    // Wait for the remaining cycle time to keep Controller loop cadence
+    void FFNetAgent::wait(void)
+    {
+        geopm_time_s current_time;
+        do {
+            geopm_time(&current_time);
+        }
+        while (geopm_time_diff(&m_last_wait, &current_time) < M_WAIT_SEC);
+        geopm_time(&m_last_wait);
+    }
+
+    // Describes expected policies to be provided by the resource manager or user
+    std::vector<std::string> FFNetAgent::policy_names(void)
+    {
+        return {"PERF_ENERGY_BIAS"};
+    }
+    // Describes samples to be provided to the resource manager or user
+    std::vector<std::string> FFNetAgent::sample_names(void)
+    {
+        return {};
+    }
+
+    // Adds the wait time to the top of the report
+    std::vector<std::pair<std::string, std::string> > FFNetAgent::report_header(void) const
+    {
+        return {{"Wait time (sec)", std::to_string(M_WAIT_SEC)}};
+    }
+
+    std::vector<std::pair<std::string, std::string> > FFNetAgent::report_host(void) const
+    {
+        std::vector<std::pair<std::string, std::string> > result;
+        return result;
+    }
+
+    // This Agent does not add any per-region details
+    std::map<uint64_t, std::vector<std::pair<std::string, std::string> > >
+            FFNetAgent::report_region(void) const
+    {
+        return {};
+    }
+
+    // Adds trace columns signals of interest
+    std::vector<std::string> FFNetAgent::trace_names(void) const
+    {
+        std::vector<std::string> tracelist;
+        for (const m_domain_key_s domain_key : m_domains) {
+            for (const std::string& trace_name : m_net_map.at(domain_key)->trace_names()) {
+                tracelist.push_back(
+                        trace_name
+                        + M_TRACE_SUFFIX.at(domain_key.type)
+                        + std::to_string(domain_key.index));
+            }
+        }
+        return tracelist;
+    }
+
+    std::vector<std::function<std::string(double)> > FFNetAgent::trace_formats(void) const
+    {
+        return {};
+    }
+
+    // Updates the trace with values for signals from this Agent
+    void FFNetAgent::trace_values(std::vector<double> &values)
+    {
+        int vidx = 0;
+        for (const auto &kv : m_net_map) {
+            std::vector<double> domain_row = kv.second->trace_values();
+            for (std::size_t idx=0; idx < domain_row.size(); ++vidx, ++idx) {
+                values[vidx] = domain_row[idx];
+            }
+        }
+    }
+
+    void FFNetAgent::enforce_policy(const std::vector<double> &policy) const
+    {
+    }
+
+    bool FFNetAgent::is_all_nan(const std::vector<double> &vec)
```

Comment:
I see that this was just copy/pasted from e.g. the FrequencyMapAgent.  Ok for now, but we should refactor both agents and move this into Helper.cpp.

Reply:
Noted